@here/harp-geometry/lib/TechniqueProcessor.ts

```diff
@@ -0,0 +1,133 @@
+     * values according to style position in file.
+     */
+    readonly renderOrder: number;
+}
+
+/**
+ * Base class that handles processing data accordingly to its assigned [[Technique]].
+ */
+export abstract class TechniqueProcessor {
+    /**
+     * [[Technique]] identifier.
+     */
+    static readonly id: string = "none";
+
+    /**
+     * Array of uniforms assigned to the objects created with this [[TechniqueProcessor]].
+     */
+    static readonly uniforms: Map<string, THREE.Uniform>;
+    /**
+     * Vertex shader source code.
+     */
+    static readonly vertexShader: string;
+    /**
+     * Fragment shader source code.
+     */
+    static readonly fragmentShader: string;
+
+    /**
+     * Checks if this [[TechniqueProcessor]] can process the specified data.
+     *
+     * @param data Data to check for support.
+     */
+    static supportsData(data: ArrayBufferLike | {}): boolean {
+        return false;
+    }
+    /**
+     * Checks if this [[TechniqueProcessor]] supports this type of decoding info.
+     *
+     * @param data Data to check for support.
+     */
+    static supportsDecodingInfo(decodingInfo: {}): boolean {
+        return false;
+    }
+
+    /**
+     * Creates the expected [[THREE.BufferGeometry]] for the specified data and the [[Technique]]
+     * associated with this [[TechniqueProcessor]].
+     *
+     * Internally calls `supportsData` and `supportsDecodingInfo` to check if data can be decoded.
+     *
+     * @param data Input data.
+     * @param decodingInfo Required decoding info.
+     * @param env Data environment variable.
+     * @param params [[TechniqueParams]] to apply to this data.
+     * @param target Optional target [[THREE.BufferGeometry]].
+     */
+    static createBuffers(
+        data: ArrayBufferLike | {},
+        decodingInfo: {},
+        env: MapEnv,
+        params: TechniqueParams,
+        target?: THREE.BufferGeometry
+    ): THREE.BufferGeometry | undefined {
+        return undefined;
+    }
+
+    /**
+     * Creates the expected array of [[THREE.Object3D]] for the specified [[THREE.BufferGeometry]]
+     * and the [[Technique]] associated with this [[TechniqueProcessor]].
+     *
+     * Handles uniform updates for the created objects through the material `onBeforeRender`
+     * callback.
+     *
+     * @param geometry Input geometry.
+     * @param params [[TechniqueParams]] to apply to this geometry.
+     * @param sceneState [[SceneState]] to apply to the generated objects.
+     */
+    static createObjects(
+        geometry: THREE.BufferGeometry,
+        params: TechniqueParams,
+        sceneState: SceneState
+    ): THREE.Object3D | THREE.Object3D[] {
+        return [];
+    }
+
+    protected static updateUniforms(
+        object: THREE.Object3D,
+        params: TechniqueParams,
+        sceneState: SceneState
+    ) {
+        object.onBeforeRender = (renderer, scene, camera, geom, mat, group) => {
```

Comment:
I understand your concerns, but without proper interpolant implementation (which might be on the way, and it's not there), we will do this anyways for almost all materials (look at #496 ). 

We cannot have runtime interpolation right now any other way, and we're not adding any overhead here, just rearranging where the logic is located (which is now hardcoded in ``TileGeometryCreator``).

Reply:
@robertoraggi, @amsgames94 

I think that we cannot do much in scope of this patch, so this one is ok.

Ultimate solution, would need `kind-of` MaterialCache, so we would reuse materials in tiles and pre-render pass over all materials in current scene to update dynamic uniforms dependent on time, pixel2world, and whatnot.

This would save us from tons of `onBeforeRender` calls, at least those that doesn't need toggle stencil buffer.

I must write-out this idea as patch for discussion.